### PR TITLE
Adding transpose_twice test

### LINKED ISSUE #558: [Ops] Transpose: Support for transpose over -3 dim when it's not tile-dim aligned

### PR DESCRIPTION
**Pull request by mtopalovicTT**
Adding transpose test

### CODE REVIEW

**Comment by rpavlovicTT** [comment]
> Fixes
> 
> From this IR
> 
> ```
> %0 = tensor.empty
> %1 = ttir.transpose(%arg1, %0)
> %2 = tensor.empty
> %3 = ttir.transpose(%0, %2)
> ```
> 
> To this IR
> 
> ```
> %1 = ttir.transpose(%arg1, %0)
> %0 = tensor.empty
> %2 = tensor.empty
> %3 = ttir.transpose(%0, %2)
> ```
> 
> Also fixes how we update results types of ops. For example if we take initial IR from above, layout pass will update %0 attr with correct memory space and tile type. It will update all of it's uses to reflect correct attribute but it won't update %1 to have the same attribute as %0 which will lead to this error:
> 
> ```
> test/ttmlir/Dialect/TTNN/transpose/transpose_twice.mlir:7:10: error: 'ttir.transpose' op expected type of operand #1 ('tensor<128x64xbf16, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<4x2x!tt.tile<32x32, bf16>, #tt.memory_space<dram>>>>') to match type of corresponding result ('tensor<128x64xbf16, #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<128x64xbf16, #tt.memory_space<system>>>>')
>     %1 = "ttir.transpose"(%arg0, %0) <{dim0 = 0 : si32, dim1 = 1 : si32, operand_constraints = [#any_device_tile, #any_device_tile]}> : (tensor<64x128xbf16>, tensor<128x64xbf16>) -> tensor<128x64xbf16>
> ```
> 
> And here is dump of whole IR before the error
> 
> ```
> #layout = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<64x128xbf16, #system>>
> #layout1 = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<4x2x!tt.tile<32x32, bf16>, #dram>>
> #layout2 = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<2x4x!tt.tile<32x32, bf16>, #dram>>
> #layout3 = #tt.layout<(d0, d1) -> (d0, d1), undef, <1x1>, memref<128x64xbf16, #system>>
> "builtin.module"() ({
>   "func.func"() <{function_type = (tensor<64x128xbf16, #layout>) -> tensor<64x128xbf16, #layout>, sym_name = "forward"}> ({
>   ^bb0(%arg0: tensor<64x128xbf16, #layout>):
>     %0 = "tensor.empty"() : () -> tensor<128x64xbf16, #layout1>
>     %1 = "tensor.empty"() : () -> tensor<64x128xbf16, #layout2>
>     %2 = "ttir.to_layout"(%arg0, %1) : (tensor<64x128xbf16, #layout>, tensor<64x128xbf16, #layout2>) -> tensor<64x128xbf16, #layout2>
>     %3 = "ttir.transpose"(%2, %0) <{dim0 = 0 : si32, dim1 = 1 : si32, operand_constraints = [#any_device_tile, #any_device_tile]}> : (tensor<64x128xbf16, #layout2>, tensor<128x64xbf16, #layout1>) -> tensor<128x64xbf16, #layout3>
>     %4 = "tensor.empty"() : () -> tensor<64x128xbf16, #layout2>
>     %5 = "ttir.transpose"(%0, %4) <{dim0 = 1 : si32, dim1 = 0 : si32, operand_constraints = [#any_device_tile, #any_device_tile]}> : (tensor<128x64xbf16, #layout1>, tensor<64x128xbf16, #layout2>) -> tensor<64x128xbf16, #layout2>
>     %6 = "tensor.empty"() : () -> tensor<64x128xbf16, #layout>
>     %7 = "ttir.to_layout"(%5, %6) : (tensor<64x128xbf16, #layout2>, tensor<64x128xbf16, #layout>) -> tensor<64x128xbf16, #layout>
>     "func.return"(%7) : (tensor<64x128xbf16, #layout>) -> ()
>   }) : () -> ()
> ```
> 
> You can see that %0 layout and %3 layout don't match.
> 
> I will experiment some more to see if we hit similar errors with different IR.
> 
> fixes #558

I don't get how is this correct ?
```
%1 = ttir.transpose(%arg1, %0)
%0 = tensor.empty
```

**Comment by mtopalovicTT** [comment]
@rpavlovicTT comment is miss leading, updated

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 1]
```diff
@@ -617,9 +625,18 @@ class TTIRLayoutDPSOperandsRewriter
...
+          continue;
+        }
+        rewriter.modifyOpInPlace(op, [&]() {
+          modified = true;
+          op->getResult(0).setType(operand.get().getType());
+        });
       }
     }
 
+    op->getParentOfType<mlir::ModuleOp>().print(llvm::errs());
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:639]: Stray print

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 2]
```diff
@@ -550,10 +550,17 @@ static std::optional<Value> createToLayoutOp(PatternRewriter &rewriter,
...
   auto desiredLayout = rewriter.getAttr<LayoutAttr>(
       ty, desiredMemorySpace, currLayout.getGrid(), desiredElementType);
+  tensor::EmptyOp exising_empty = input.getDefiningOp<tensor::EmptyOp>();
+
+  // If we are changing the layout of existing op
+  // we need to set the insertion point to the existing op
+  if (exising_empty) {
+    rewriter.setInsertionPoint(input.getDefiningOp());
+  }
+
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:560]: I think it's safe, and potentially more correct, to just unconditionally `setInsertionPoint`:
```c++
rewriter.setInsertionPoint(input.getDefiningOp());
auto output = rewriter.create<tensor::EmptyOp>(....
```

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 3]
```diff
@@ -608,6 +615,7 @@ class TTIRLayoutDPSOperandsRewriter
           createToLayoutOp(rewriter, op.getLoc(), operand.get(),
                            operandConstraint, defaultMemorySpace);
 
+      // If we changed layout (memory space or element type), update the op
       if (desiredLayout) {
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:619]: What's confusing to me, is why `desiredLayout` is false in this case?  We just changed layout right?

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 2 above)
Comment (mtopalovicTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:560]: I think problem is when we want to layout function argument, getDefiningOp returns nullptr

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 1 above)
Comment (mtopalovicTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:639]: Yea, left comment above to remove it :D

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 3 above)
Comment (mtopalovicTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:619]:
```
%0 = tensor.empty
%1 = ttir.transpose(%arg1, %0)
%2 = tensor.empty
%3 = ttir.transpose(%0, %2)
```

From what I observed this happens:

1. Rewrite is called on `%3 = ttir.transpose(%0, %2)`
2. We update %0 layout (this updates all references of %0 including `%1 = ttir.transpose(%arg1, %0)`
3. Later on rewrite is called for `%1 = ttir.transpose(%arg1, %0)`
4. We call createLayout for %0 operand, but it's already set to layout from step 2 so we return `nullopt` and we skip update of return type

Comment (nsmithtt) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:619]: Ok wait, sorry I just realized that I think there is a bug in the test.  The above isn't legal, the test should read:

```mlir
%0 = tensor.empty
%1 = ttir.transpose(%arg1, %0)
%2 = tensor.empty
%3 = ttir.transpose(%1, %2)
```

The _result_ (not the output arg) from DPS ops must be used in order to properly express the dependency between ops.  At the TTIR graph level, we should treat all values as though they are copied, i.e. `%0` is an argument to transpose that gets copied to `%1` and then "thrown away".  During lowering we can apply additional semantics, like move semantics from C++, where we can say that the contents from `%0` are immediately moved into `%1`.  A lowering pass that properly models tensor allocation can prove this and leverage memory aliasing, but from a graph level we need to abide by DPS semantics.

Relevant section of [MLIR docs](https://mlir.llvm.org/docs/Bufferization/#destination-passing-style):

> `outs` operands are referred to as “destinations” in the following (quotes are important as this operand isn’t modified in place but copied) and comes into place in the context of bufferization as a possible “anchor” for the bufferization algorithm. This allows the user to shape the input in a form that guarantees close to optimal bufferization result when carefully choosing the SSA value used as “destination”.
>
> For every tensor result, a DPS op has a corresponding tensor operand. If there aren’t any other conflicting uses of this tensor, the bufferization can alias it with the op result and perform the operation “in-place” by reusing the buffer allocated for this “destination” input.

**Review by nsmithtt — CHANGES_REQUESTED** [review]
See inline comment, let's fix the test to see if the issue still persists.  Sorry for not noticing this sooner!

File: lib/Dialect/TTIR/Transforms/Passes.cpp (hunk 3 above)
Comment (mtopalovicTT) [reply lib/Dialect/TTIR/Transforms/Passes.cpp:619]: I see. Ok will check tmrw when I fix the test...

**Comment by mtopalovicTT** [comment]
@nsmithtt Fixed bugs... Thanks again for pointing out

File: lib/Dialect/TTIR/Transforms/Passes.cpp [hunk 4]
```diff
@@ -606,10 +606,16 @@ createToLayoutOp(PatternRewriter &rewriter, Location loc, Value input,
   auto desiredLayout =
       rewriter.getAttr<LayoutAttr>(ty, desiredMemorySpace, currLayout.getGrid(),
                                    desiredElementType, desiredMemLayout);
+  tensor::EmptyOp exising_empty = input.getDefiningOp<tensor::EmptyOp>();
+  // If we are changing the layout of existing op
+  // we need to set the insertion point to the existing op
+  if (exising_empty) {
+    rewriter.setInsertionPoint(input.getDefiningOp());
+  }
```
Comment (nsmithtt) [inline lib/Dialect/TTIR/Transforms/Passes.cpp:614]: So I'm thinking now that this `setInsertionPoint` might not be necessary anymore.  The insertion point will have already been set to be the current op we're rewriting so things should "just work".

**Comment by mtopalovicTT** [comment]
Need repro from FE team for resnet transpose which didn't work....

**Review by nsmithtt — APPROVED** [review]
Thanks!

**Review by sdjordjevicTT — APPROVED** [review]

**Pull request merged**